Diagnose missing expr in non-void single-expr func. [52025782]

nate-chandler opened the pull request:
After SE-0255, compiling

```
func foo() -> Int {
    return
}
```

would result in a diagnostic without source location:

```
<unknown>:0: error: cannot convert return expression of type '()' to
```

Now, it results in

```
filename.swift:8:5: error: non-void function should return a value
    return
    ^
```

as it did prior to SE-0255.

To achieve that, during type checking for statements, when the StmtChecker visits return statements, check whether we are within a non-void-returning, single-expression function and that that single expression is an implicit empty tuple.  (An empty implicit tuple is added as the result of a resultless return statement that appears as the only element in a single-expression function.)

To facilitate that, `hasSingleExpressionBody` and `getSingleExpressionBody` was added to `AnyFunctionRef` and added `getSingleExpressionBody` to `AbstractClosureExpr` (which already had `hasSingleExpressionBody`).

rdar://problem/52025782

nate-chandler committed:
Diagnose missing expr in non-void single-expr func. [52025782]

After SE-0255, compiling

```
func foo() -> Int {
    return
}
```

would result in a diagnostic without source location:

```
<unknown>:0: error: cannot convert return expression of type '()' to
```

Now, it results in

```
filename.swift:8:5: error: non-void function should return a value
    return
    ^
```

as it did prior to SE-0255.

To achieve that, during type checking for statements, when the
StmtChecker visits return statements, check whether we are within a
non-void-returning, single-expression function and that that single
expression is an implicit empty tuple.  (An empty implicit tuple is
added as the result of a resultless return statement that appears as the
only element in a single-expression function.)

To facilitate that, `hasSingleExpressionBody` and
`getSingleExpressionBody` was added to `AnyFunctionRef` and added
`getSingleExpressionBody` to `AbstractClosureExpr` (which already had
`hasSingleExpressionBody`).

rdar://problem/52025782

nate-chandler commented:
@swift-ci Please smoke test

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -442,6 +442,32 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
+      auto expr = TheFunc->getSingleExpressionBody();
+      if (expr->isImplicit()) {
+        if (auto tuple = dyn_cast<TupleExpr>(expr)) {
+          if (tuple->getNumElements() == 0 && !ResultTy->isVoid()) {
```
You already checked `ResultTy->isVoid()` at the top level. (It might also make sense to merge this with the previous `if`, since they're checking related conditions.)

nate-chandler replied:
Fixed in https://github.com/apple/swift/pull/25802/commits/3761b02d74721cf0ccedae16ba17c2500ef5637a .  Is 

```
if (expr->isImplicit() &&
  dyn_cast<TupleExpr>(expr) != nullptr &&
  dyn_cast<TupleExpr>(expr)->getNumElements() == 0) {
```

what you had in mind for merging the ifs together?

jrose-apple replied:
I actually meant

```
if (ResultTy->isVoid()) {
  if (!RS->hasResult()) { /* old stuff */ }
  if (TheFunc->hasSingleExpressionBody()) { /* new stuff */ }
}
```

Up to you, though.

nate-chandler replied:
If I'm understanding this right, that wouldn't quite work because the old stuff

```
if (!RS->hasResult()) {
  if (!ResultTy->isVoid())
    TC.diagnose(RS->getReturnLoc(), diag::return_expr_missing);
  return RS;
}
```

returns `RS` always whenever `!RS->hasResult()` so there'd have to be an additional check after the `if (ResultTy->isVoid()) {...}` for `if (!RS->hasResult())`.

jrose-apple replied:
Oops, you're right. I didn't notice that the return wasn't nested.

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -442,6 +442,32 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
+        if (auto tuple = dyn_cast<TupleExpr>(expr)) {
+          if (tuple->getNumElements() == 0 && !ResultTy->isVoid()) {
+            TC.diagnose(RS->getReturnLoc(), diag::return_expr_missing);
+            return nullptr;
```
I think you can still return `RS` here, like the other case does. Or does that cause more failures?

nate-chandler replied:
I thought it result in the earlier location-less failure staying in, but I was confused.  Fixed in https://github.com/apple/swift/pull/25802/commits/3761b02d74721cf0ccedae16ba17c2500ef5637a .

jrose-apple reviewed (commented):
Glad it worked out nicely!

nate-chandler committed:
Fixes based on PR comments.

- Avoided extraneous rechecking void-ness of ResultTy.
- Returned RS rather than nullptr in the diagnosed case.

nate-chandler commented:
@swift-ci Please smoke test

jrose-apple commented on lib/Sema/TypeCheckStmt.cpp:
```diff
@@ -442,6 +442,30 @@ class StmtChecker : public StmtVisitor<StmtChecker, Stmt*> {
+      auto expr = TheFunc->getSingleExpressionBody();
+      if (expr->isImplicit() &&
+          dyn_cast<TupleExpr>(expr) != nullptr &&
+          dyn_cast<TupleExpr>(expr)->getNumElements() == 0) {
```
This isn't what I meant, but I think it works. You can simplify to `isa<TupleExpr>(expr) && cast<TupleExpr>(expr)->`, though.

nate-chandler replied:
That is much nicer!  Fixed in https://github.com/apple/swift/pull/25802/commits/bdf1c0b243978103b58b026c3a90edfbd934f80f .

jrose-apple replied:
`cast` was correct, by the way. It's an asserting version of `dyn_cast`, but since you just checked with `isa` it's fine. (I guess it'll be optimized either way, but it's still convention.)

nate-chandler replied:
Oops.  I meant to commit that change (as the commit message shows) but failed to actually commit it.  Fixed in https://github.com/apple/swift/pull/25802/commits/6dab25bb3e39010aea76a8f6f5b236a8f83f0a5f .  And thank you for the explanation!

nate-chandler committed:
Fixes based on PR comments.

- Replaced double dyn_cast with isa and cast.

nate-chandler commented:
@swift-ci Please smoke test

nate-chandler committed:
Fixes based on PR comments.

- Actually replaced dyn_cast cast.

nate-chandler commented:
@swift-ci Please smoke test

nate-chandler commented:
@swift-ci Please clean smoke test Linux platform

nate-chandler commented:
@swift-ci Please clean smoke test Linux platform

nate-chandler commented:
@swift-ci Please clean smoke test Linux platform

shahmishal commented:
@swift-ci Please smoke test Linux
